File: backend/app/schemas/project_snapshots.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


SnapshotType = Literal["validation", "report"]


class ProjectSnapshotBase(BaseModel):
    project_id: str
    snapshot_type: SnapshotType
    name: str
    notes: str | None = None
    created_by: str | None = None


class ProjectSnapshotSummaryResponse(ProjectSnapshotBase):
    id: str
    created_at: datetime

    model_config = {"from_attributes": True}


class CreateProjectSnapshotRequest(BaseModel):
    snapshot_type: SnapshotType
    name: str | None = Field(default=None, max_length=255)
    notes: str | None = None
    created_by: str | None = Field(default=None, max_length=128)


class ProjectSnapshotDetailResponse(ProjectSnapshotSummaryResponse):
    payload: dict[str, Any]


class SnapshotComparisonDelta(BaseModel):
    current: float
    snapshot: float
    delta: float


class ProjectSnapshotComparisonResponse(BaseModel):
    snapshot: ProjectSnapshotSummaryResponse
    snapshot_type: SnapshotType
    deltas: dict[str, SnapshotComparisonDelta]

File: frontend/src/types/projectSnapshots.ts
import type {
  ProjectReportSummaryResponse,
  ProjectValidationSummaryResponse,
} from "./projectViews";

export type SnapshotType = "validation" | "report";

export type ProjectSnapshotSummary = {
  id: string;
  project_id: string;
  snapshot_type: SnapshotType;
  name: string;
  notes: string | null;
  created_by: string | null;
  created_at: string;
};

export type ProjectSnapshotDetail<TPayload = unknown> = ProjectSnapshotSummary & {
  payload: TPayload;
};

export type SnapshotComparisonDelta = {
  current: number;
  snapshot: number;
  delta: number;
};

export type ProjectSnapshotComparison = {
  snapshot: ProjectSnapshotSummary;
  snapshot_type: SnapshotType;
  deltas: Record<string, SnapshotComparisonDelta>;
};

export type CreateProjectSnapshotRequest = {
  snapshot_type: SnapshotType;
  name?: string;
  notes?: string;
  created_by?: string;
};

export type ValidationSnapshotDetail = ProjectSnapshotDetail<ProjectValidationSummaryResponse>;
export type ReportSnapshotDetail = ProjectSnapshotDetail<ProjectReportSummaryResponse>;
